akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala

```diff
@@ -2285,6 +2285,286 @@ private[akka] final class StatefulMap[S, In, Out](
+        case Success(elemOpt) =>
+          elemOpt match {
+            case Some(elem) => emit(out, elem, () => failOrCompleteStage(ex))
+            case None       => failOrCompleteStage(ex)
+          }
+        case Failure(releaseEx) => failStage(releaseEx)
+      }
+
+      private val releaseCB =
+        getAsyncCallback[(Option[Throwable], Try[Option[Out]])]((emitThenFailOrCompleteStage _).tupled).invoke _
+
+      private def releaseThenFailStage(ex: Throwable): Unit = {
+        if (stateAcquired) {
+          state.foreach(s => {
+            try {
+              val future = onComplete(s)
+              stateAcquired = false
+              future.value match {
+                case None =>
+                  future.onComplete(v => releaseCB(Some(ex) -> v))(akka.dispatch.ExecutionContexts.parasitic)
+                case Some(v) =>
+                  emitThenFailOrCompleteStage(Some(ex), v)
+              }
+            } catch {
+              // ideally should never throw when release resource
+              case NonFatal(e) => failStage(e)
+            }
+          })
+        }
+      }
+
+      private def releaseThenCompleteStage(): Unit = {
+        if (stateAcquired) {
+          state.foreach(s => {
+            try {
+              val future = onComplete(s)
+              stateAcquired = false
+              future.value match {
+                case None =>
+                  future.onComplete(getAsyncCallback[Try[Option[Out]]] {
+                    case Failure(e) => failStage(e)
+                    case Success(output) =>
+                      output match {
+                        case Some(v) => emit(out, v, () => initState())
+                        case None    => initState()
+                      }
+                  }.invoke)(akka.dispatch.ExecutionContexts.parasitic)
+                case Some(v) =>
+                  emitThenFailOrCompleteStage(None, v)
+              }
+            } catch {
+              // ideally should never throw when release resource
+              case NonFatal(e) => failStage(e)
+            }
+          })
+        }
+      }
+
+      override def onUpstreamFinish(): Unit =
+        if (buffer.isEmpty && earlyPulledElem.isEmpty) {
+          releaseThenCompleteStage()
+        }
+
+      override def onUpstreamFailure(ex: Throwable): Unit = if (state.isDefined) {
+        releaseThenFailStage(ex)
+      }
+
+      override def onDownstreamFinish(cause: Throwable): Unit = if (state.isDefined) {
+        releaseThenFailStage(cause)
+      }
+
+      override def postStop(): Unit = {
+        state match {
+          case Some(value) =>
+            // outlet isn't available here
+            if (stateAcquired) onComplete(value)
+          //TODO figure out what should be done here
+          // since resource maybe in the half way of acquisition
+          case None =>
+        }
+      }
+
+      override def preStart(): Unit = {
```

Comment:
Could initState be invoked after the first pull, not on preStart?

Reply:
I personally don't hold any opinion whether state should be initialized lazily or not, but this operator is intent to implement *ResourceAsync API, and by that means the state here is more likely a resource thus it's more appropriate to give user a clear view that when the resource be acquired especially when this operator doesn't integrate any extra lifecycle manage mechanism like refreshing IMO.

Reply:
I have created an issue about this:https://github.com/akka/akka/issues/31609